### DISCUSSION TITLE
Автозапуск скрипта на MacOS [РЕШЕНИЕ]

### DISCUSSION THREAD

**Discussion by lovipomidorku** [xvzc/SpoofDPI#115] [Show and tell]
### Скрипт для запуска spoof-dpi на MacOS

1. Сохраняем файл spoof-dpi.sh в любое удобное место. _(Например, на рабочий стол для быстрого доступа)_

2. В свойствах файла меняем приложение для открытия по умолчанию на "Терминал"

3. В Терминале пишем команду `chmod +x ` и перетягиваем файл spoof-dpi.sh в Терминал.

4. Открываем файл как исполняемый. Profit! 

[spoof-dpi.sh.zip](https://github.com/user-attachments/files/16525322/spoof-dpi.sh.zip)

**Comment by asimba** [comment]
А кто мешает прописать в сам spoof-dpi в automator и добавить в объекты входа.

**Reply by lovipomidorku** [reply]
Кто мешает прописать и выложить скрипт автоматора?

**Reply by asimba** [reply]
Тот же кто и мешает dmg/pkg сделать - маков мало и лень. А автоматор на живой железке - процедура минут на 5.

**Reply by lovipomidorku** [reply]
> Тот же кто и мешает dmg/pkg сделать - маков мало и лень. А автоматор на живой железке - процедура минут на 5.

Делать самому лень, а критиковать мое решение не лень. Странно 🤨

**Reply by asimba** [reply]
Агась. :) Приношу свои извинения за столь неуклюжую попытку получить готовое решение в стиле macOS с кнопкой "сделать хорошо" (для "казуалов"). Могу поделиться такими решениями в виде установочных пакетов для Windows и Debian/Ubuntu (ибо их написать было не лень, таких пользователей у меня больше).

**Reply by lovipomidorku** [reply]
> Агась. :) Приношу свои извинения за столь неуклюжую попытку получить готовое решение в стиле macOS с кнопкой "сделать хорошо" (для "казуалов"). Могу поделиться такими решениями в виде установочных пакетов для Windows и Debian/Ubuntu (ибо их написать было не лень, таких пользователей у меня больше).

Не сомневаюсь в вашей компетенции, но пинать мое решение доставляет вам удовольствие?)

**Reply by asimba** [reply]
Ничуть. Каюсь лишь в том, что хотел сподвигнуть Вас на большее (после сравнительной оценки затрат времени на объяснение конечному пользователю macOS последовательности необходимых действий и того сколько я потратил в конечном итоге на полную сборку (что не было не столько необходимо, сколько самую малость интересно), и прописывание в Automator одной единственной строки).

**Comment by 00enterausername00** [comment]
Альтернативный вариант:

В папке ~/Library/LaunchAgents/ создаем файл с названием org.spoof-dpi.plist 
В него помещаем следующее содержимое:

```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>Label</key>
	<string>org.spoof-dpi.plist</string>
	<key>ProgramArguments</key>
	<array>
		<string>/Users/YOUUSER/.spoof-dpi/bin/spoof-dpi</string>
	</array>
	<key>RunAtLoad</key>
	<true/>
	<key>KeepAlive</key>
    <true/>
</dict>
</plist>
```

Меняем YOUUSER на ваше имя пользователя, сохраняем файл.
Теперь программа будет автоматически запускаться при каждой загрузке системы без каких-либо дополнительных окон.

Для активации службы без перезагрузки в терминале вводим (предварительно убедившись что у вас не запущена другая копия программы): `launchctl load -w ~/Library/LaunchAgents/org.spoof-dpi.plist`
Если нужно убить программу на время текущей сессии, вводим в терминале `launchctl unload ~/Library/LaunchAgents/org.spoof-dpi.plist`

Этот фрагмент:
```
<key>KeepAlive</key>
    <true/>
```
отвечает за автоматический перезапуск службы, с ним вы не сможете убить службу через мониторинг системы. Его можно удалить, если такое поведение вас не устраивает.

В следующем сообщении попробую выложить файл для автоматора, чтобы можно было запускать/убивать программу мышкой

**Reply by 00enterausername00** [reply]
Предполагается, что файл размещен в папке автоматора в айклауд драйв. Если по понятным причинам вы не хотите скачивать исполняемые файлы от ноунеймов из интернета, то идете в автоматор, нажимаете создать программу, добавляете блок "запустить shell скрипт" и вставляете следующий код:
```
if [ -z $(launchctl list | grep spoof) ];
then
plutil -replace CFBundleIconFile -string "figa-on" ~/Library/Mobile\ Documents/com\~apple\~Automator/Documents/DPI.app/Contents/Info.plist
touch ~/Library/Mobile\ Documents/com\~apple\~Automator/Documents/DPI.app;
rm /var/folders/*/*/*/com.apple.dock.iconcache;
killall Dock; 
launchctl load -w ~/Library/LaunchAgents/org.spoof-dpi.plist;
else
plutil -replace CFBundleIconFile -string "figa-off" ~/Library/Mobile\ Documents/com\~apple\~Automator/Documents/DPI.app/Contents/Info.plist
touch ~/Library/Mobile\ Documents/com\~apple\~Automator/Documents/DPI.app;
rm /var/folders/*/*/*/com.apple.dock.iconcache;
killall Dock; 
launchctl unload ~/Library/LaunchAgents/org.spoof-dpi.plist;
fi
```
Данный код проверяет запущена ли служба, запускает либо убивает ее в зависимости от результата из первого пункта, меняет иконку приложения, чистит кэш иконок дока и перезагружает док. Если не предполагается размещать ярлык в доке или не нужна меняющаяся в зависимости от статуса службы иконка, то соответствующие пункты можно удалить. 

### **Будьте внимательны! Эта программа будет работать только если вы создали службу по инструкции из моего предыдущего сообщения, оставили название файла DPI.app и поместили эту программу в папку /Library/Mobile\ Documents/com\~apple\~Automator/Documents** 

Буду очень признателен, если кто-то подскажет способ, позволяющий не перезагружать док, но менять при этом внешний вид иконки в доке.

[DPI.app.zip](https://github.com/user-attachments/files/16646540/DPI.app.zip)

**Comment by tollfree** [comment]
Есть [GUI программа](https://github.com/SpoofDPIApp/SpoofDPI-App/releases/latest) для запуска

**Reply by nklsnkls** [reply]
только от макоси 13-ой, на монтерее не идет(

**Comment by z4ffe** [comment]
Легче всего через homebrew
